include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -517,20 +552,52 @@ MultigridPreconditionerBase<dim, Number>::do_initialize_dof_handler_and_constrai
     // setup dof-handler and constrained dofs for each p-level
     for(auto level : p_levels)
     {
-      // setup dof_handler: create dof_handler...
+      // setup dof_handler
       auto dof_handler = new dealii::DoFHandler<dim>(*tria);
-      // ... create FE and distribute it
+
+      // distribute dofs
       if(level.is_dg)
+      {
         dof_handler->distribute_dofs(
           dealii::FESystem<dim>(dealii::FE_DGQ<dim>(level.degree), n_components));
+      }
       else
+      {
         dof_handler->distribute_dofs(
           dealii::FESystem<dim>(dealii::FE_Q<dim>(level.degree), n_components));
+      }
+
+      // distribute MG dofs
       dof_handler->distribute_mg_dofs();
-      // setup constrained dofs:
+
+      // constrained dofs
       auto constrained_dofs = new dealii::MGConstrainedDoFs();
       constrained_dofs->clear();
-      this->initialize_constrained_dofs(*dof_handler, *constrained_dofs, dirichlet_bc);
+
+      if(level.is_dg)
+      {
+        std::set<dealii::types::boundary_id> dirichlet_boundary;
+        for(auto & it : dirichlet_bc)
+          dirichlet_boundary.insert(it.first);
+        constrained_dofs->initialize(*dof_handler);
+        constrained_dofs->make_zero_boundary_constraints(*dof_handler, dirichlet_boundary);
+      }
+      else
+      {
+        constrained_dofs->initialize(*dof_handler);
+        for(auto it : dirichlet_bc)
+        {
+          std::set<dealii::types::boundary_id> dirichlet_boundary;
+          dirichlet_boundary.insert(it.first);
+
+          dealii::ComponentMask mask    = dealii::ComponentMask();
+          auto                  it_mask = dirichlet_bc_component_mask.find(it.first);
+          if(it_mask != dirichlet_bc_component_mask.end())
+            mask = it_mask->second;
+
+          constrained_dofs->make_zero_boundary_constraints(*dof_handler, dirichlet_boundary, mask);
+        }
```

Comment:
I will comment on this also in the general section, but I feel that this might be a performance concern if we have cases with a high number of boundary ids. The reason is how the underlying implementation in deal.II works by a loop over all mesh elements. Thus, using a single `std::set` of boundary ids for all cases where the component mask is the same would be preferable. (I worked hard to make loops over mesh elements in deal.II faster during the last year, and I invite others to make further improvements, but the underlying complexity cannot be avoided.)